### PR TITLE
Blueprint https://blueprints.launchpad.net/or/+spec/activity-events-t…

### PR DESCRIPTION
**Pull request by Csantucci**
…riggered-by-ai-trains

### CODE REVIEW

**Comment by Csantucci** [comment]
Request to James to merge into the master OpenRails files modified for blueprint https://blueprints.launchpad.net/or/+spec/activity-events-triggered-by-ai-trains . Seven source files to be modified.

**Comment by Csantucci** [comment]
Request to James to merge into the master OpenRails files modified for blueprint https://blueprints.launchpad.net/or/+spec/activity-events-triggered-by-ai-trains . Seven source files to be modified.

File: Source/Orts.Formats.Msts/ActivityFile.cs [hunk 1]
```diff
@@ -1133,6 +1157,60 @@ public ORTSWeatherChange(STFReader stf)
...
+                            case "Ground":
+                                ORTSActSoundFileType = 3;
+                                break;
+                            case "Location":
+                                ORTSActSoundFileType = 4;
+                                break;
+                            default:
+                                ORTSActSoundFileType = 0;
+                                break;
+                        }
```
Comment (twpol) [inline Source/Orts.Formats.Msts/ActivityFile.cs]: This looks like it should be an enum, as then you can skip the whole switch and just parse straight into the value.

File: Source/Orts.Simulation/Simulation/Activity.cs [hunk 2]
```diff
@@ -1502,17 +1518,23 @@ override public Boolean Triggered(Activity activity)
         {
             var triggered = false;
             var e = this.ParsedObject as Orts.Formats.Msts.EventCategoryLocation;
+            Train train = Simulator.PlayerLocomotive.Train;
```
Comment (twpol) [inline Source/Orts.Simulation/Simulation/Activity.cs]:
```suggestion
            var train = Simulator.PlayerLocomotive.Train;
```

File: Source/Orts.Simulation/Simulation/Activity.cs [hunk 3]
```diff
@@ -1502,17 +1518,23 @@ override public Boolean Triggered(Activity activity)
...
+            if (ParsedObject.TrainService != "" && Train != null)
+            {
+                if (Train.FrontTDBTraveller != null) train = Train;
+                else return triggered;
```
Comment (twpol) [inline Source/Orts.Simulation/Simulation/Activity.cs]: Swapping this around so the return is first might make this more readable, as you'll not need the `else`.

File: Source/RunActivity/Viewer3D/Sound.cs [hunk 4]
```diff
@@ -2630,8 +2646,22 @@ public void Update()
                             }
                             break;
                         case 3:
-                            var worldLocation = Program.Viewer.Simulator.PlayerLocomotive.WorldPosition.WorldLocation;
+                            var loco = (train == Program.Viewer.Simulator.PlayerLocomotive.Train) ?
+                                Program.Viewer.Simulator.PlayerLocomotive : train.Cars[0];
+                            var worldLocation = loco.WorldPosition.WorldLocation;
                             worldLocation.Location.Y = worldLocation.Location.Y + 3; // Sound does not come from earth!
+//                            string wsName = Program.Viewer.Simulator.RoutePath + @"\WORLD\" + WorldFile.WorldFileNameFromTileCoordinates(worldLocation.TileX, worldLocation.TileZ) + "s";
```
Comment (twpol) [inline Source/RunActivity/Viewer3D/Sound.cs]:
```suggestion
```

File: Source/RunActivity/Viewer3D/Sound.cs [hunk 5]
```diff
@@ -2630,8 +2646,22 @@ public void Update()
...
+                            Program.Viewer.SoundProcess.AddSoundSources(localEventID, new List<SoundSourceBase>() { ActivitySounds });
+                            break;
+                        case 4:
+                            worldLocation = WorldLocation.None;
+                            worldLocation.TileX = activitySound.TileX;
+                            worldLocation.TileZ = activitySound.TileZ;
+                            worldLocation.Location.X = activitySound.X;
+                            worldLocation.Location.Y = activitySound.Y + 3; // Sound does not come from earth!
+                            worldLocation.Location.Z = activitySound.Z;
+ //                           wsName = Program.Viewer.Simulator.RoutePath + @"\WORLD\" + WorldFile.WorldFileNameFromTileCoordinates(worldLocation.TileX, worldLocation.TileZ) + "s";
```
Comment (twpol) [inline Source/RunActivity/Viewer3D/Sound.cs]:
```suggestion
```

File: Source/RunActivity/Viewer3D/Sound.cs [hunk 6]
```diff
@@ -2660,8 +2690,21 @@ public void Update()
                             }
                             break;
                         case 3:
-                            var worldLocation = Program.Viewer.Simulator.PlayerLocomotive.WorldPosition.WorldLocation;
+                            var loco = (train == Program.Viewer.Simulator.PlayerLocomotive.Train) ?
+                                Program.Viewer.Simulator.PlayerLocomotive : train.Cars[0];
+                            var worldLocation = loco.WorldPosition.WorldLocation;
                             worldLocation.Location.Y = worldLocation.Location.Y + 3; // Sound does not come from earth!
+ //                           string wsName = Program.Viewer.Simulator.RoutePath + @"\WORLD\" + WorldFile.WorldFileNameFromTileCoordinates(worldLocation.TileX, worldLocation.TileZ) + "s";
```
Comment (twpol) [inline Source/RunActivity/Viewer3D/Sound.cs]:
```suggestion
```

File: Source/RunActivity/Viewer3D/Sound.cs [hunk 7]
```diff
@@ -2660,8 +2690,21 @@ public void Update()
...
+                            ActivitySounds = new SoundSource(Program.Viewer, worldLocation, Events.Source.None, ORTSActSoundFile, true, ORTSActSoundFileType, true);
+                            Program.Viewer.SoundProcess.AddSoundSources(localEventID, new List<SoundSourceBase>() { ActivitySounds });
+                            break;
+                        case 4:
+                            worldLocation = WorldLocation.None;
+                            worldLocation.TileX = activitySound.TileX;
+                            worldLocation.TileZ = activitySound.TileZ;
+                            worldLocation.Location.X = activitySound.X;
+                            worldLocation.Location.Y = activitySound.Y + 3; // Sound does not come from earth!
+//                            wsName = Program.Viewer.Simulator.RoutePath + @"\WORLD\" + WorldFile.WorldFileNameFromTileCoordinates(worldLocation.TileX, worldLocation.TileZ) + "s";
```
Comment (twpol) [inline Source/RunActivity/Viewer3D/Sound.cs]:
```suggestion
```

File: Source/Orts.Simulation/Simulation/Activity.cs (hunk 2 above)
Comment (Csantucci) [reply Source/Orts.Simulation/Simulation/Activity.cs]: OK. Should I click on "Commit suggestion"? Do I have then to make the same change in my local and remote copy to keep everything aligned?

File: Source/Orts.Simulation/Simulation/Activity.cs (hunk 3 above)
Comment (Csantucci) [reply Source/Orts.Simulation/Simulation/Activity.cs]: Will do.

File: Source/Orts.Simulation/Simulation/Activity.cs (hunk 2 above)
Comment (twpol) [reply Source/Orts.Simulation/Simulation/Activity.cs]: I believe the "Commit suggestion" will modify the branch in GitHub with my suggested change, so afterwards you will need to do `git pull` to get the change locally.

File: Source/Orts.Formats.Msts/ActivityFile.cs (hunk 1 above)
Comment (twpol) [reply Source/Orts.Formats.Msts/ActivityFile.cs]: Yeah, that looks like the kind of thing I mean regarding enum.

If you make the change locally and push it on the same branch as this PR (strongly recommended) the PR will update automatically, including emailing me about it.

**Comment by Csantucci** [comment]
James, I'm stuck with the official way. My remote repository does not accept my push, because it has been changed by the small commits that I have done here (deletion of commented lines). If I try to make a pull from my remote repository into my local repository, it says they are aligned... So I pass here to you the compound patch file of my changes as requested by you, hoping that you can use this for the merge.

[ActivityEvents_twpol.zip](https://github.com/openrails/openrails/files/2803903/ActivityEvents_twpol.zip)

**Comment by Csantucci** [comment]
Correction to above: if I try to make a pull from my remote repositori into my local repository, it generates merge conflicts and there I'm not able to handle them.

**Comment by Csantucci** [comment]
OK, I think I was able to do the merge and to push it.

**Comment by twpol** [comment]
Thanks for persevering, it looks like the changes are all there. I'll review the code tomorrow as it's late here already.

File: Source/Orts.Formats.Msts/ActivityFile.cs [hunk 8]
```diff
@@ -846,25 +855,15 @@ public override void AddOrModifyEvent (STFReader stf, string fileName)
...
-                                ORTSActSoundFileType = 0;
-                                break;
-	                    }
+                    try
+                    {
+                    ORTSActSoundFileType = (ORTSActSoundFileTypes)Enum.Parse(typeof(ORTSActSoundFileTypes), stf.ReadString());
+                    }
+                    catch(ArgumentException)
+                    {
+                        STFException.TraceInformation(stf, "Skipped unknown activity sound file type " + stf.ReadString());
```
Comment (twpol) [inline Source/Orts.Formats.Msts/ActivityFile.cs]: I'm not 100% sure but will this show the correct thing in the message?

**Pull request merged**